### PR TITLE
Add support for creating semi-minimal kernels for BESSPIN

### PR DESCRIPTION
**Pull request by brooksdavis**

### CODE REVIEW

**Comment by brooksdavis** [comment]
oops, this branch has a couple local changes that I'd committed and never pushed. Should be fairly harmless.

File: pycheribuild/mtree.py [hunk 1]
```diff
@@ -255,6 +256,29 @@ def __contains__(self, item):
         mtree_path = self._ensure_mtree_path_fmt(str(item))
         return mtree_path in self._mtree
 
+    def filter(self, globs, exceptions=[], print_status=False):
```
Comment (jrtc27) [inline pycheribuild/mtree.py:259]:
```suggestion
    def filter_out(self, globs, exceptions=[], print_status=False):
```
? This to me implies positive not negative filtering.

File: pycheribuild/projects/cross/openssl.py [hunk 2]
```diff
@@ -51,16 +50,23 @@ def setup(self):
         self.configure_args.append(str(self.source_dir / "Configure"))
         self.configure_args.append("BSD-generic64")
         self.configure_args.append("-shared")
-        self.configure_args.append("--install-prefix=" + str(self.destdir))
+	# XXX: should be based on version
```
Comment (jrtc27) [inline pycheribuild/projects/cross/openssl.py:53]: Spaces

File: pycheribuild/projects/cross/openssl.py [hunk 3]
```diff
@@ -51,16 +50,23 @@ def setup(self):
...
+        if (self.project_name != "fett-openssl"):
+            self.configure_args.append("--prefix=" + str(self.destdir) + str(self._install_prefix))
```
Comment (jrtc27) [inline pycheribuild/projects/cross/openssl.py:55]: No parens, and looking at `self.project_name` is ugly

File: pycheribuild/projects/disk_image.py [hunk 4]
```diff
@@ -1069,6 +1069,236 @@ def cheribsd_class(self):
...
+    include_boot = True
+
+    @classmethod
+    def setup_config_options(cls, **kwargs):
+        super().setup_config_options(default_hostname=_default_disk_image_hostname("cheribsd-besspin"),
+                                     extra_files_suffix="-besspin", **kwargs)
+        cls.strip_binaries = cls.add_bool_option(
+            "strip", default=True, help="strip ELF files to reduce size of generated image")
+        cls.use_cheribsd_purecap_rootfs = cls.add_bool_option(
+            "use-cheribsd-purecap-rootfs", default=True, help="Use the rootfs built by cheribsd-purecap instead")
```
Comment (jrtc27) [inline pycheribuild/projects/disk_image.py:1085]: Do we really need this? It's a silly target to build, just leave it as hybrid and thus uninteresting?

File: pycheribuild/projects/cross/openssl.py (hunk 3 above)
Comment (brooksdavis) [reply pycheribuild/projects/cross/openssl.py:55]: I'm hoping to purge the FETT bits in October...

File: pycheribuild/projects/disk_image.py (hunk 4 above)
Comment (brooksdavis) [reply pycheribuild/projects/disk_image.py:1085]: I'll remove this. It's leftover from foking the minimal target.

Comment (jrtc27) [reply pycheribuild/projects/disk_image.py:1085]: ... we should probably remove that for the same reason (though at least that one is default=False)

File: pycheribuild/mtree.py (hunk 1 above)
Comment (brooksdavis) [reply pycheribuild/mtree.py:259]: Thinking this morning, I wonder if `exclude_matching` would be better (it had been `filter_matching` initially in my tree)

File: pycheribuild/projects/disk_image.py [hunk 5]
```diff
@@ -1069,6 +1069,223 @@ def cheribsd_class(self):
...
+            "libasn1.so.11",
+            "libypclnt.so.4",
+            "libcom_err.so.5",
+            "libhx509.so.11",
+            "libwind.so.11",
+            "libheimbase.so.11",
+            "libprivateheimipcc.so.11",
+            "libssl.so.111"
+        ]
+        self.mtree.exclude_matching("./usr/lib/*", list(map(lambda p: "./usr/lib/" + p, usr_lib_exceptions)))
```
Comment (arichardson) [inline pycheribuild/projects/disk_image.py:1244]: Maybe better to just keep the `pam_*` exclusion and add the libraries explicitly afterwards? That way we should get a build error on a version bump/library rename.

File: pycheribuild/projects/disk_image.py [hunk 6]
```diff
@@ -1069,6 +1069,223 @@ def cheribsd_class(self):
...
+        self.mtree.exclude_matching("./usr/libexec/*", ["*/getty"])
+
+        usr_sbin_exceptions = [
+            "service",
+            "sshd",
+            "newsyslog",
+            "utx",
+            "syslogd"
+        ]
+        self.mtree.exclude_matching("./usr/sbin/*", list(map(lambda p: "./usr/sbin/" + p, usr_sbin_exceptions)))
```
Comment (arichardson) [inline pycheribuild/projects/disk_image.py:1255]: Much less of a problem here, but maybe removing all and then explicitly adding the 5 binaries is makes this code easier to understand? Some of the arrays are positive, others are negative.

**Review by arichardson — APPROVED** [review]
Could we use this to build minimal FreeBSD images (since we can't use cheribsdbox there)?

**Pull request merged**